[Docs][torch.compile] NNCF quantization/compression (#27934)
openvinotoolkit/openvino: docs/articles_en/openvino-workflow/torch-compile.rst

       if sys.version_info >= (3, 11):
           `raise RuntimeError("Python 3.11+ not yet supported for torch.compile")

TorchServe Integration
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

TorchServe is a performant, flexible, and easy to use tool for serving PyTorch models in production. For more information on the details of TorchServe,
you can refer to `TorchServe github repository. <https://github.com/pytorch/serve>`__. With OpenVINO ``torch.compile`` integration into TorchServe you can serve
PyTorch models in production and accelerate them with OpenVINO on various Intel hardware. Detailed instructions on how to use OpenVINO with TorchServe are

> available in `TorchServe examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__

Reviewer: Hi @daniil-lyakhov : I was thinking here, as we mention about general TorchServe examples..

> available in `TorchServe examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__

Reviewer: ```suggestion
available in `TorchServe examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__ and in a `use case app <https://github.com/pytorch/serve/tree/master/examples/usecases/llm_diffusion_serving_app>__`.
```

Support for Automatic1111 Stable Diffusion WebUI
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Automatic1111 Stable Diffusion WebUI is an open-source repository that hosts a browser-based interface for the Stable Diffusion
based image generation. It allows users to create realistic and creative images from text prompts.
Stable Diffusion WebUI is supported on Intel CPUs, Intel integrated GPUs, and Intel discrete GPUs by leveraging OpenVINO
``torch.compile`` capability. Detailed instructions are available in
`Stable Diffusion WebUI repository. <https://github.com/openvinotoolkit/stable-diffusion-webui/wiki/Installation-on-Intel-Silicon>`__

Model Quantization and Compression
######################################

> 

Reviewer 2: I would suggest to add some introduction here and cover the following topics:
- OpenVINO backend in torch.compile supports 
  - NNCF TorchFX Quantization and Compression
  - PyTorch 2 Export Quantization
- Advantages of  NNCF TorchFX Quantization and compression comparing with PyTorch 2 Export Quantization 
- NNCF is the default way to compress models for the OpenVINO backend in torch.compile.

Author: Done, please check

NNCF Model Optimization Support (Preview)
+++++++++++++++++++++++++++++++++++++++++++++

Neural Network Compression Framework(`NNCF <https://github.com/daniil-lyakhov/nncf/tree/develop/examples/post_training_quantization/torch_fx/resnet18>`__)

> implements advanced quantization and compression algorihtms which could be applied to ``torch.fx.GraphModule`` to speed up inference

Reviewer 3: ```suggestion
implements advanced quantization and compression algorithms which could be applied to ``torch.fx.GraphModule`` to speed up inference
```

and decrease memory consumption:

> .. code-block:: python

Reviewer 2: I would suggest to make separate examples for quantization and compression.

Author: Done, please check

   import nncf
   import openvino.torch
   import torch

   calibration_loader = torch.utils.data.DataLoader(...)

   def transform_fn(data_item):
       images, _ = data_item
       return images

   # Model quantization
   quantized_model = nncf.quantize(model, calibration_dataset)
   # or compression
   # compressed_model = nncf.compress_model(model)

   quantized_model = torch.compile(quantized_model, backend="openvino")
   # compressed_model = torch.compile(compressed_model, backend="openvino")

NNCF unlocks the full potential of low-precision OpenVINO kernels due to quantizers placement designed specifically for the OpenVINO.

> Advanced algorithms like ``SmoothQuant`` or ``BiasCorrection`` allows furher metrics improvement minimizing the outputs discrepancy between the original and compressed models.

Reviewer 3: ```suggestion
Advanced algorithms like ``SmoothQuant`` or ``BiasCorrection`` allows further metrics improvement minimizing the outputs discrepancy between the original and compressed models.
```

For further details, please see the `documentation <https://docs.openvino.ai/2024/openvino-workflow/model-optimization.html>`__

> and a `tutorial <https://github.com/daniil-lyakhov/nncf/tree/develop/examples/post_training_quantization/torch_fx/resnet18>`__.

Reviewer 3: I think link is incorrect. it should be official main repo

Author: My bad, fixed, thanks!

PyTorch models in production and accelerate them with OpenVINO on various Intel hardware. Detailed instructions on how to use OpenVINO with TorchServe are
available in `TorchServe examples. <https://github.com/pytorch/serve/tree/master/examples/pt2/torch_compile_openvino>`__ and in a `use case app <https://github.com/pytorch/serve/tree/master/examples/usecases/llm_diffusion_serving_app>`__.

Support for Automatic1111 Stable Diffusion WebUI
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Automatic1111 Stable Diffusion WebUI is an open-source repository that hosts a browser-based interface for the Stable Diffusion
based image generation. It allows users to create realistic and creative images from text prompts.
Stable Diffusion WebUI is supported on Intel CPUs, Intel integrated GPUs, and Intel discrete GPUs by leveraging OpenVINO
``torch.compile`` capability. Detailed instructions are available in
`Stable Diffusion WebUI repository. <https://github.com/openvinotoolkit/stable-diffusion-webui/wiki/Installation-on-Intel-Silicon>`__

Model Quantization and Compression
######################################

> Model quantization and compression allows to speed up inference speed and decrease memory consumption at the cost of minor model outputs deviations.
> `torch.compile` OpenVINO backend supports two model optimization APIs:

Reviewer 2: ```suggestion
Model quantization and compression are effective methods for accelerating model inference and reducing memory consumption, with minimal impact on model accuracy. The `torch.compile` OpenVINO backend supports two key model optimization APIs:
```

> 1. Neural Network Compression Framework(`NNCF <https://github.com/daniil-lyakhov/nncf/tree/develop/examples/post_training_quantization/torch_fx/resnet18>`__).

Reviewer 2: Please add a cross link to NNCF OpenVINO documentation instead of link to the example.

Author: Done

> 2. PyTorch 2 export quantization.

Reviewer 2: May be add some description about NNCF and PyTorch 2 Export Quantization. For example:
1. Neural Network Compression Framework (NNCF)
NNCF offers advanced algorithms for post-training quantization and compression in OpenVINO toolkit.
2. PyTorch 2 Export Quantization
A general-purpose API designed for quantizing models captured by `torch.export`

Author: Done, please check

> NNCF is the preferred way of model quantization and compression. While general PyTorch 2 export quantization API indeed decreases model latency,
> NNCF is tuned to achieve optimal results specifically for the OpenVINO backend in terms of inference speed and model metrics.

Reviewer 2: ```suggestion
NNCF is the recommended approach for model quantization and compression. NNCF specifically optimizes the model for the OpenVINO backend, providing optimal results in terms of inference speed and accuracy.
```

Author: the model -> models

+++++++++++++++++++++++++++++++++++++++++++++

Neural Network Compression Framework(`NNCF <https://github.com/daniil-lyakhov/nncf/tree/develop/examples/post_training_quantization/torch_fx/resnet18>`__)
implements advanced quantization and compression algorihtms which could be applied to ``torch.fx.GraphModule`` to speed up inference
and decrease memory consumption:

.. code-block:: python

   import nncf
   import openvino.torch
   import torch

   calibration_loader = torch.utils.data.DataLoader(...)

   def transform_fn(data_item):
       images, _ = data_item
       return images

   # Model quantization
   quantized_model = nncf.quantize(model, calibration_dataset)
   # or compression
   # compressed_model = nncf.compress_model(model)

   quantized_model = torch.compile(quantized_model, backend="openvino")
   # compressed_model = torch.compile(compressed_model, backend="openvino")

For further details, please see the `documentation <https://docs.openvino.ai/2024/openvino-workflow/model-optimization.html>`__
and a `tutorial <https://github.com/daniil-lyakhov/nncf/tree/develop/examples/post_training_quantization/torch_fx/resnet18>`__.

TorchServe Integration

> +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Reviewer: We can also add this as a usecase sample: https://github.com/pytorch/serve/tree/master/examples/usecases/llm_diffusion_serving_app

Author: Hi @ravi9, I could not find a reference to the nncf usage in your example, perhaps we need to put this link somewhere else
